draft-fossati-tls-attestation.md

```diff
@@ -377,11 +386,74 @@ The Certificate payload is used as a container, as shown in
           CertificateEntry certificate_list<0..2^24-1>;
       } Certificate;
 ~~~~
-{: #figure-certificate title="Certificate Message."}
+{: #figure-attest-only title="Certificate Message when using only attestation."}
 
 The encoding of the evidence structure is defined in
 {{I-D.ftbs-rats-msg-wrap}}.
 
+## Attestation alongside X.509 certificates
+
+When the chosen evidence type indicates usage of both attestation and PKIX, the
+X.509 certificate will serve as the main payload in the Certificate message,
+while the attestation evidence will be carried in the CertificateEntry
+extension, as shown in {{figure-cert-attest}}.
+
+~~~~
+      struct {
+          select (certificate_type) {
+              case RawPublicKey:
+                /* From RFC 7250 ASN.1_subjectPublicKeyInfo */
+                opaque ASN1_subjectPublicKeyInfo<1..2^24-1>;
+
+              case attestation:
+                opaque evidence<1..2^24-1>;
+              
+              /* X.509 certificate conveyed as usual */
+              case X509:
+                opaque cert_data<1..2^24-1>;
+          };
+          /* attestation evidence conveyed as an extension */
+          Extension extensions<0..2^16-1>;
+      } CertificateEntry;
+
+      struct {
+          opaque certificate_request_context<0..2^8-1>;
+          CertificateEntry certificate_list<0..2^24-1>;
+      } Certificate;
+~~~~
+{: #figure-cert attest title="Certificate Message when using PKIX and attestation."}
+
+The encoding of the evidence structure is defined in
+{{I-D.ftbs-rats-msg-wrap}}.
+
+As described in Appendix A, this authentication mechanism is meant primarily for
+carrying platform attestation evidence to provide more context to the relying
+party. This evidence must be cryptographically bound to the TLS handshake to
+prevent relay attacks. A Binder Collection as described in Appendix B is
+therefore used when the attestation scheme does not allow the binding data to be
+part of the token. The structure of the collection is given in
+{{figure-tls-binder}}.
+
+~~~~
+binder_collection = {
+  &(nonce: 1) => bstr .size (8..64)
+  &(ik_pub_fingerprint: 2) => bstr .size 32
+  &(channel_binder: 3) => bstr .size 32
+}
+~~~~
+{: #figure-tls-binder title="Format of TLS Binder Collection."}
+
+* Nonce is the value provided as a challenge by the relying party.
+* The identity key public fingerprint (ik_pub_fingerprint) is a SHA256 hash of
+  the Subject Public Key Info from the leaf X.509 certificate transmitted in the
+  handshake.
+* The channel binder (channel_binder) is a partial transcript of the TLS
+  handshake, up to (but not including) the Certificate message.
+
+A SHA256 digest of the encoded binder must be included in the attestation
```

Comment:
Any good reason why we're using a hardcoded hash algorithm rather than the hash alg negotiated by TLS?

Reply:
I'm not that familiar with TLS, so wasn't aware this was an option, I simply wanted to avoid conveying the hash algorithms used for this as well. Will change to use that instead.